src/placement/jump_map.c

```diff
@@ -1019,6 +1118,84 @@ jump_map_obj_find_reint(struct pl_map *map, struct daos_obj_md *md,
 	return rc < 0 ? rc : idx;
 }
 
+static int
+jump_map_obj_find_addition(struct pl_map *map, struct daos_obj_md *md,
+			   struct daos_obj_shard_md *shard_md,
+			uint32_t reint_ver, uint32_t *tgt_rank,
+			uint32_t *shard_id, unsigned int array_size,
+			int myrank)
+{
+	struct pl_jump_map              *jmap;
+	struct pl_obj_layout            *layout;
+	struct pl_obj_layout            *add_layout;
+	d_list_t                        remap_list;
+	d_list_t                        add_list;
+	struct jm_obj_placement         jop;
+	int                             rc;
+
+	int idx = 0;
+
+	D_DEBUG(DB_PL, "Finding new layout for server addition\n");
+
+	/* Caller should guarantee the pl_map is up-to-date */
+	if (pl_map_version(map) < reint_ver) {
+		D_ERROR("pl_map version(%u) < rebuild version(%u)\n",
+			pl_map_version(map), reint_ver);
+		return -DER_INVAL;
+	}
+
+	jmap = pl_map2jmap(map);
+
+	rc = jm_obj_placement_get(jmap, md, shard_md, &jop);
+	if (rc) {
+		D_ERROR("jm_obj_placement_get failed, rc %d.\n", rc);
+		return rc;
+	}
+
+	/* Allocate space to hold the layout */
+	rc = pl_obj_layout_alloc(jop.jmop_grp_size, jop.jmop_grp_nr, &layout);
+	if (rc)
+		return 0;
+	rc = pl_obj_layout_alloc(jop.jmop_grp_size, jop.jmop_grp_nr,
+				 &add_layout);
+	if (rc)
+		goto out;
+
+	D_INIT_LIST_HEAD(&remap_list);
+	D_INIT_LIST_HEAD(&add_list);
```

Comment:
These INIT_LIST should be done before goto out, otherwise it might cause problem in the following remap_list_free_all in the error handling path.